extensions-core/druid-pac4j/src/main/java/org/apache/druid/security/pac4j/JwtAuthFilter.java

```diff
@@ -0,0 +1,125 @@
+import com.nimbusds.jwt.JWTParser;
+import com.nimbusds.openid.connect.sdk.claims.IDTokenClaimsSet;
+import org.apache.druid.java.util.common.logger.Logger;
+import org.apache.druid.server.security.AuthConfig;
+import org.apache.druid.server.security.AuthenticationResult;
+import org.pac4j.core.context.HttpConstants;
+import org.pac4j.oidc.profile.creator.TokenValidator;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import java.io.IOException;
+import java.text.ParseException;
+import java.util.Optional;
+
+public class JwtAuthFilter implements Filter
+{
+  private static final Logger LOG = new Logger(JwtAuthFilter.class);
+
+  private final String authorizerName;
+  private final String name;
+  private final OIDCConfig oidcConfig;
+  private final TokenValidator tokenValidator;
+
+  public JwtAuthFilter(String authorizerName, String name, OIDCConfig oidcConfig, TokenValidator tokenValidator)
+  {
+    this.authorizerName = authorizerName;
+    this.name = name;
+    this.oidcConfig = oidcConfig;
+    this.tokenValidator = tokenValidator;
+  }
+
+  @Override
+  public void init(FilterConfig filterConfig)
+  {
+
+  }
+
+  @Override
+  public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain filterChain)
+      throws IOException, ServletException
+  {
+    // Skip this filter if the request has already been authenticated
+    if (servletRequest.getAttribute(AuthConfig.DRUID_AUTHENTICATION_RESULT) != null) {
+      filterChain.doFilter(servletRequest, servletResponse);
+      return;
+    }
+
+    HttpServletRequest httpServletRequest = (HttpServletRequest) servletRequest;
+    Optional<String> idToken = extractBearerToken(httpServletRequest);
+
+    if (idToken.isPresent()) {
+      try {
+        // Parses the JWT and performs the ID Token validation specified in the OpenID spec: https://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation
+        IDTokenClaimsSet claims = tokenValidator.validate(JWTParser.parse(idToken.get()), null);
+        if (claims != null) {
+          Optional<String> claim = Optional.of(claims.getStringClaim(oidcConfig.getOidcClaim()));
+
+          if (claim.isPresent()) {
+            LOG.debug("Authentication successful for " + oidcConfig.getClientID());
+            AuthenticationResult authenticationResult = new AuthenticationResult(
+                claim.get(),
+                authorizerName,
+                name,
+                null
+            );
+            servletRequest.setAttribute(AuthConfig.DRUID_AUTHENTICATION_RESULT, authenticationResult);
+          } else {
+            LOG.error(
+                "Authentication failed! Please ensure the validity of the ID token and contains the configured claim.");
```

Comment:
```suggestion
                "Authentication failed! Please ensure that ID token is valid and it contains the configured claim.");
```

Reply:
My initial thought was to not break the authentication chain in case the user has any other custom authenticators further down the chain. But i see that if the request already has an id token, they have signaled intent to use that to authenticate.